core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

```diff
@@ -152,28 +155,25 @@ private[spark] class MetricsSystem private (
     } else { defaultName }
   }
 
-  def getSourcesByName(sourceName: String): Seq[Source] = sources.synchronized {
-    sources.filter(_.sourceName == sourceName).toSeq
+  def getSourcesByName(sourceName: String): Seq[Source] = sourcesWithListeners.synchronized {
+    sourcesWithListeners.keySet.filter(_.sourceName == sourceName).toSeq
   }
 
-  def registerSource(source: Source): Unit = {
-    sources.synchronized {
-      sources += source
-    }
+  def registerSource(source: Source): Unit = sourcesWithListeners.synchronized {
     try {
-      val regName = buildRegistryName(source)
-      registry.register(regName, source.metricRegistry)
+      val listener = new MetricsSystemListener(buildRegistryName(source))
+      source.metricRegistry.addListener(listener)
+      sourcesWithListeners += source -> listener
     } catch {
```

Comment:
Let us keep the scope of the lock narrow to match what was there earlier, unless there is a good reason to expand the critical section ?